clang/test/SemaSYCL/deferred-diagnostics-emit.cpp

```diff
@@ -0,0 +1,88 @@
+// RUN: %clang_cc1 -I %S/Inputs -fsycl -triple spir64 -fsycl-is-device -verify -fsyntax-only %s
+//
+// Ensure that the SYCL diagnostics that are typically deferred, correctly emitted.
+//
+
+#include <sycl.hpp>
+
+struct S {
+  virtual void foo() {}
+};
+
+int calledFromKernel(int a){
+  // expected-error@+1 {{zero-length arrays are not permitted in C++}}
+  int MalArray[0];  
+
+  // expected-error@+1 {{__float128 is not supported on this target}}
+  __float128 malFloat = 40;  
+
+  // not sure if 'no virtual function' is a _deferred_ diagnostic, testing anyway 
+  S mal;
+  // expected-error@+1 {{SYCL kernel cannot call a virtual function}}
+  mal.foo();  
+
+  return a + 20;
+}
+
+
+//  template used to specialize a function that contains a lambda that should
+//  result in a deferred diagnostic being emitted.
+//  HOWEVER, this is not working presently.  
+//  TODO: re-test after new deferred diagnostic system is merged. 
+//        restore the "FIX!!" tests below
+
+template <typename T>
+void setup_sycl_operation(const T VA[]) {
+    
+  cl::sycl::range<1> numOfItems;
+  cl::sycl::queue deviceQueue;
+
+  deviceQueue.submit([&](cl::sycl::handler &cgh) {
+    cgh.single_task<class AName>([=]() {
+      // FIX!!  xpected-error@+1 {{zero-length arrays are not permitted in C++}}
+      int OverlookedBadArray[0]; 
+                       
+      // FIX!!   xpected-error@+1 {{__float128 is not supported on this target}}
+      __float128 overlookedBadFloat = 40; 
+ 
+    });
+  });
+}
+
+int main(int argc, char **argv) {
+
+  // --- direct lambda testing ---
+  cl::sycl::range<1> numOfItems;
+  cl::sycl::queue deviceQueue;
+
+  
+  deviceQueue.submit([&](cl::sycl::handler &cgh) {
+  
+    cgh.single_task<class AName>([=]() {
```

Comment:
@Fznamznon  
Can you help with with this test? I have one last "note" I need to suppress.

 When running `llvm-lit -v deferred-diagnostics-emit.cpp`  I get this:
```
error: 'note' diagnostics seen but not expected: 
  File /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp Line 182: called by 'kernel_single_task<AName, (lambda at /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp:61:34)>'
  File /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp Line 182: called by 'kernel_single_task<AName, (lambda at /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp:61:34)>'
```

So it looks like I should add this expectation:
```
// expected-note@+1 2{{called by 'kernel_single_task<AName, (lambda}}
```
But no matter where I put that expectation it never works. First it tells me it expected the note but didn't see it, then it says it saw the note without expecting it.  Awesome.

```
error: 'note' diagnostics expected but not seen: 
  File /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp Line 61 (directive at /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp:60): called by 'kernel_single_task<AName, (lambda
  File /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp Line 61 (directive at /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp:60): called by 'kernel_single_task<AName, (lambda
error: 'note' diagnostics seen but not expected: 
  File /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp Line 182: called by 'kernel_single_task<AName, (lambda at /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp:61:34)>'
  File /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/Inputs/sycl.hpp Line 182: called by 'kernel_single_task<AName, (lambda at /iusers/cperkins/sycl_workspace/llvm/clang/test/SemaSYCL/deferred-diagnostics-emit.cpp:61:34)>'
```

Is the expectation correct?    I haven't been able to find docs for them, so I'm just cribbing from other places.

I tried putting the expectation at line 58, line 60, in the argument list of `cgh.single_task` so at to precede the lambda, line 62, line 76, etc.  But nothing works.

Reply:
Is there a compiler flag to suppress the notes?

Reply:
First, I don't think that we should suppress notes in tests.
Second, I see what happens. You are trying to add note expectation to wrong place.
See logs, the emitted note points to function named `kernel_single_task` and it's defined in file test/SemaSYCL/Inputs/sycl.hpp on line 182.
By the construction like `// expected-note@+1 2{{called by 'kernel_single_task<AName, (lambda}}` in `test/SemaSYCL/deferred-diagnostics-emit.cpp` file on line `60` means that you expect `2` same notes with this text in this file on line `61`.
The trick here that you use `single_task` function from our fake sycl runtime, and `kernel_single_task` call happens inside it. I'm not sure that I know how to expect notes from other file (maybe @erichkeane knows). As a fix you can declare your own `kernel_single_task` function in your test file and note expectation will point to correct file.

Reply:
Ok - I re-introduced the bespoke SYCL namespace code into the test and put the expectation into it. Now the llvm-lit passes and the test correctly stresses the namespace aspect we require.